File: proxies/bind/bind.pl
#! /usr/bin/perl

use Time::HiRes qw(sleep);
use IO::Socket::INET;

# Initialize Values
$lhost = "{{LHOST}}";
$lport = {{LPORT}};
$rhost = "{{RHOST}}";
$rport = {{RPORT}};
$data_size = 4096;
$verbose = 1;
$client_read = 1;

sub start {
    # Start proxy
    my $s = _bind();
    if ($verbose) {
        print "Listening on $lhost:$lport\n";
    }
    while (1) {
        my $conn = $s->accept();
        if ($verbose) {
            my $ip = $conn->peerhost();
            my $port = $conn->peerport();
            print "Connection received from $ip:$port\n";
        }
        _switch($conn);
    }
}

sub _switch {
    # Switch between read/write
    my $client_sock = $_[0];
    $client_sock->blocking(0); # Non-blocking
    my $server_sock = _connect();
    $server_sock->blocking(0); # Non-blocking
    my $received;
    while (1) {
        # Read from client
        if ($client_read) {
            $received = _recv($client_sock);
            if (!defined($received)) {
                last;
            }
            elsif ($received) {
                _sendall($server_sock, $received);
            }
            else {
                $client_read = 0;
                sleep(0.005);
            }
        }
        # Read from server
        if (!$client_read) {
            $received = _recv($server_sock);
            if (!defined($received)) {
                last;
            }
            elsif ($received) {
                _sendall($client_sock, $received);
            }
            else {
                $client_read = 1;
                sleep(0.005);
            }
        }
    }
    # Close the sockets if not already closed
    _close($client_sock, $server_sock);
}

sub _recv {
    # Receive bytes
    my $recv_sock = $_[0];
    my $data = "";
    my $bytes_read;
    if (defined($bytes_read = $recv_sock->recv($data, $data_size))) {
        if (!$bytes_read and !$data) {
            # Client disconnected gracefully
            return undef;
        }
        return $data;
    }
    return "";
}

sub _sendall {
    my $sock = $_[0];
    my $data_to_send = $_[1];
    my $total_sent = 0;
    my $data_length = length($data_to_send);
    while ($total_sent < $data_length) {
        my $bytes_sent = $sock->send(substr($data_to_send, $total_sent), 0);
        if (!defined $bytes_sent) {
            die "Error sending data: $!";
        }
        $total_sent += $bytes_sent;
    }
}

sub _close {
    # This function is not used in the main flow now but kept for completeness
    my $client = $_[0];
    my $server = $_[1];
    $client->close();
    $server->close();
    if ($verbose) {
        print "Proxied connection closed!\n";
    }
}

sub _connect {
    # Connect to the server
    my $s = new IO::Socket::INET(
        PeerAddr => $rhost,
        PeerPort => $rport,
        Proto    => 'tcp',
        Type     => SOCK_STREAM,
    ) or die "Failed to connect to $rhost:$rport\n";
    if ($verbose) {
        print "Connected to remote host on $rhost:$rport\n";
    }
    return $s;
}

sub _bind {
    my $s = IO::Socket::INET->new(
        LocalAddr => $lhost,
        LocalPort => $lport,
        Proto     => 'tcp',
        Listen    => 25, 
        ReuseAddr => 1, 
    ) or die "Cannot bind listener on $lhost:$lport!\n";
    return $s;
}

start();

File: proxies/bind/bind.rb
#! /usr/bin/env ruby

require 'socket'

class BindProxy

  def initialize(lhost, lport, rhost, rport, verbose=true)
    # Initialize TCP proxy
    @lhost = lhost
    @lport = lport
    @rhost = rhost
    @rport = rport
    @data_size = 4096
    @verbose = verbose
  end

  def start
    s = _bind
    if s
      s.listen(25)
      if @verbose
        puts "Listening on #{@lhost}:#{@lport}"
      end
      while true
        begin
          conn, addr = s.accept
          if @verbose
            ip = addr.ip_address
            port = addr.ip_port
            puts "Connection received from #{ip}:#{port}"
          end
          _switch(conn)
        rescue Interrupt
          break
        rescue => e
          puts "ERROR: #{e.message}"
          break
        end
      end
    end
  end

  def _switch(client_sock)
    server_sock = _connect
    if client_sock && server_sock
      client_read = true
      while true
        # Recv from client
        if client_read
          begin
            data = client_sock.recv_nonblock(@data_size)
            if data
              _sendall(server_sock, data)
            else
              break
            end
          rescue IO::WaitReadable
            client_read = false
            sleep 0.005
          rescue
            break
          end
        end
        # Recv from server
        if !client_read
          begin
            data = server_sock.recv_nonblock(@data_size)
            if data
              _sendall(client_sock, data)
            else
              break
            end
          rescue IO::WaitReadable
            client_read = true
            sleep 0.005
          rescue
            break
          end
        end
      end
      # Close the sockets if any are still open
      _close(client_sock, server_sock)
    end
  end

  def _sendall(sock, data)
    bytes_sent = 0
    while bytes_sent < data.length
      sent_count = sock.write(data.byteslice(bytes_sent..-1))
      bytes_sent += sent_count
    end
  end

  def _close(client, server)
    # Close client and server sockets
    begin
      client.close
    rescue
      nil
    end
    begin
      server.close
    rescue
      nil
    end
    if @verbose
      puts "Proxied connection closed!"
    end
  end

  def _connect
    # Connect to the server
    begin
      s = Socket.new(Socket::AF_INET, Socket::SOCK_STREAM, 0)
      addr = Socket.pack_sockaddr_in(@rport, @rhost)
      s.connect(addr)
      if @verbose
        puts "Connected to remote host on #{@rhost}:#{@rport}"
      end
      return s
    rescue
      return nil
    end
  end

  def _bind
    begin
      s = Socket.new(Socket::AF_INET, Socket::SOCK_STREAM, 0)
      s.setsockopt(Socket::SOL_SOCKET, Socket::SO_REUSEADDR, 1)
      addr = Addrinfo.tcp(@lhost, @lport)
      s.bind(addr)
      return s
    rescue
      return nil
    end
  end
end

bndprx = BindProxy.new("{{LHOST}}", {{LPORT}}, "{{RHOST}}", {{RPORT}})
bndprx.start
